pkg/types/dsse/v0.0.1/entry.go

```diff
@@ -0,0 +1,420 @@
+		Value:     swag.String(hex.EncodeToString(payloadHash[:])),
+	}
+
+	envelopeHash := sha256.Sum256([]byte(*dsseObj.ProposedContent.Envelope))
+	dsseObj.EnvelopeHash = &models.DSSEV001SchemaEnvelopeHash{
+		Algorithm: swag.String(models.DSSEV001SchemaEnvelopeHashAlgorithmSha256),
+		Value:     swag.String(hex.EncodeToString(envelopeHash[:])),
+	}
+
+	// we've gotten through all processing without error, now update the object we're unmarshalling into
+	v.DSSEObj = *dsseObj
+	v.env = env
+
+	return nil
+}
+
+// Canonicalize returns a JSON representation of the entry to be persisted into the log. This
+// will be further canonicalized by JSON Canonicalization Scheme (JCS) before being written.
+//
+// This function should not use v.DSSEObj.ProposedContent fields as they are client provided and
+// should not be trusted; the other fields at the top level are only set server side.
+func (v *V001Entry) Canonicalize(_ context.Context) ([]byte, error) {
+	canonicalEntry := models.DSSEV001Schema{
+		Signatures:      v.DSSEObj.Signatures,
+		EnvelopeHash:    v.DSSEObj.EnvelopeHash,
+		PayloadHash:     v.DSSEObj.PayloadHash,
+		ProposedContent: nil, // this is explicitly done as we don't want to canonicalize the envelope
+	}
+
+	sort.Slice(canonicalEntry.Signatures, func(i, j int) bool {
+		return *canonicalEntry.Signatures[i].Signature < *canonicalEntry.Signatures[j].Signature
+	})
+
+	itObj := models.DSSE{}
+	itObj.APIVersion = swag.String(APIVERSION)
+	itObj.Spec = &canonicalEntry
+
+	return json.Marshal(&itObj)
+}
+
+// AttestationKey and AttestationKeyValue are not implemented so the envelopes will not be persisted in Rekor
+
+func (v V001Entry) CreateFromArtifactProperties(_ context.Context, props types.ArtifactProperties) (models.ProposedEntry, error) {
+	returnVal := models.DSSE{}
+	re := V001Entry{
+		DSSEObj: models.DSSEV001Schema{
+			ProposedContent: &models.DSSEV001SchemaProposedContent{},
+		},
+	}
+	var err error
+	artifactBytes := props.ArtifactBytes
+	if artifactBytes == nil {
+		if props.ArtifactPath == nil {
+			return nil, errors.New("path to artifact file must be specified")
+		}
+		if props.ArtifactPath.IsAbs() {
+			return nil, errors.New("dsse envelopes cannot be fetched over HTTP(S)")
+		}
+		artifactBytes, err = os.ReadFile(filepath.Clean(props.ArtifactPath.Path))
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	env := &dsse.Envelope{}
+	if err := json.Unmarshal(artifactBytes, env); err != nil {
+		return nil, fmt.Errorf("payload must be a valid DSSE envelope: %w", err)
+	}
+
+	allPubKeyBytes := make([][]byte, 0)
+	if len(props.PublicKeyBytes) > 0 {
+		allPubKeyBytes = append(allPubKeyBytes, props.PublicKeyBytes...)
+	}
+
+	if len(props.PublicKeyPaths) > 0 {
+		for _, path := range props.PublicKeyPaths {
+			if path.IsAbs() {
+				return nil, errors.New("dsse public keys cannot be fetched over HTTP(S)")
+			}
+
+			publicKeyBytes, err := os.ReadFile(filepath.Clean(path.Path))
```

Comment:
Is there any threat of SSRF? path is controlled by the caller. Reading a random file should cause a failure on verification though

Reply:
this is only invoked from the CLI process so no SSRF concern